tests/Test/Xrefcheck/TooManyRequestsSpec.hs

```diff
@@ -149,35 +158,118 @@ test_tooManyRequests = testGroup "429 response tests"
+          timeoutTestCase [Delay] False
+            (cNetworkingL . ncMaxTimeoutRetriesL .~ 2)
+      , testCase "Returns an error if there were 429 but no timeouts allowed" $
+          timeoutTestCase [Respond429, Delay] False
+            (cNetworkingL . ncMaxTimeoutRetriesL .~ 0)
+      , testCase "Succeeds if there were 429 and one timeout allowed" $
+          timeoutTestCase [Respond429, Delay] True
+            (cNetworkingL . ncMaxTimeoutRetriesL .~ 1)
+      , testCase "Fails on second timeout if there were 429 and one timeout allowed" $
+          timeoutTestCase [Respond429, Delay, Delay] False
+            (cNetworkingL . ncMaxTimeoutRetriesL .~ 1)
+      , testCase "Fails on second timeout if there were 429 and one timeout allowed" $
+          timeoutTestCase [Respond429, Delay, Delay] False
+            (cNetworkingL . ncMaxTimeoutRetriesL .~ 1)
+      , testCase "Fails on maximum allowed errors achieved (mixed errors)" $
+          timeoutTestCase [Respond429, Delay, Respond429, Delay] False $ \c -> c
+            & cNetworkingL . ncMaxTimeoutRetriesL .~ 3
+            & cNetworkingL . ncMaxRetriesL .~ 3
+      , testCase "Fails on timeout if another domain returned 429" $ do
+          setRef <- newIORef S.empty
+          checkMultipleLinksWithServer
+            (mockTimeout (sec 0.4) [Respond429, Ok, Delay])
+            setRef
+            [ VerifyLinkTestEntry{ vlteConfigModifier = \c -> c
+                                     & cNetworkingL . ncMaxRetriesL .~ 1
+                                     & setAllowedTimeout
+                                 , vlteLink = "http://127.0.0.1:5000/timeout"
+                                 , vlteProgress = mkProgressWithOneTask True
+                                 , vlteExpectation = VerifyResult []
+                                 }
+            , VerifyLinkTestEntry{ vlteConfigModifier = \c -> c
+                                     & cNetworkingL . ncMaxTimeoutRetriesL .~ 0
+                                     & setAllowedTimeout
+                                 , vlteLink = "http://localhost:5000/timeout"
+                                 , vlteProgress = mkProgressWithOneTask False
+                                 , vlteExpectation = VerifyResult
+                                   [ ExternalHttpTimeout (Just $ DomainName "localhost")
+                                   ]
+                                 }
+            ]
+      ]
+
   ]
   where
-    checkLinkAndProgressWithServer mock link progress vrExpectation =
+    setAllowedTimeout = cNetworkingL . ncExternalRefCheckTimeoutL .~ (sec 0.25)
+
+    mkProgressWithOneTask shouldSucceed =
+                  Progress{ pTotal = 1
+                          , pCurrent = 1
+                          , pErrorsUnfixable = if shouldSucceed then 0 else 1
+                          , pErrorsFixable = 0
+                          , pTaskTimestamp = Nothing
+                          }
```

Comment:
Btw, I would try to avoid code indented like this.

I know this codebase has such code and I suppose you likely just followed it, but for the new code let's try to put `{` at the separate line. E.g. something like this works

```hs
    mkProgressWithOneTask shouldSucceed =
      Progress
        { pTotal = 1
        , pCurrent = 1
        , pErrorsUnfixable = if shouldSucceed then 0 else 1
        , pErrorsFixable = 0
        , pTaskTimestamp = Nothing
        }
```

or this

```hs
   mkProgressWithOneTask shouldSucceed = Progress
     { pTotal = 1
     , ...
     }
```

This does not apply to one-liners ofc, there indentation is not a concern.

No need to refactor everything in the project according to this rule, maybe only in a separate PR. For now let's follow this rule only for the new code.